morsecode.c: store characters in-place

ghuysmans opened the pull request:
Hi,

It looks like you don't need to store the decoded characters in a separate place, you can store `NUL` to signal invalid sequences and check that the index is still within bounds.

ghuysmans committed:
morsecode.c: store characters in-place

skeeto commented:
Thanks for the patch, Guillaume! I'm happy you shared this. However, this 
decoder does not quite have the behavior I want, and I've added a new test 
(9880128) to illustrate. The original tests could also be more accurate,
detecting that invalid inputs are rejected as early as possible, not just 
at the terminator. For instance ".-.-" should reject the second dash and 
not wait for the terminator. My original decoder would pass such a test, 
but not this new decoder.

Immediate rejections are important: (1) allows callers to pinpoint the bad 
input, and (2) prevents the state from overflowing. Addressing (2) on its 
own isn't difficult. Return zero instead of a state beyond -63.

Thinking about how to fix (1) without re-introducing edge information, I 
don't like the answers. It's insufficient to peek ahead one level because 
there are "dead" spots within decode paths, such as at 4 dashes, which is 
not a valid stopping point, but may continue to '0' and '9'. It would need 
to traverse two levels in search of a valid path. It would also need to 
either bounds check that search, or extend the tree with zeroes, so that 
everything below the last level is at least a virtual zero.

So while the edge information bits are technically redundant, it's a small 
price to pay so that this information need not be resolved at run time via 
traversal.

ghuysmans committed:
morsecode.c: Test invalid character handling

ghuysmans committed:
morsecode.c: Restore early bound check

ghuysmans commented:
Thanks for your detailed reply. I tested it first without digits, I should've been more careful.

We might think that a bound check and a special (but inelegant) test for `.-.-` are enough, but I think we can still do better: we don't need to traverse the tree every time, we can compute statically whether the current strict prefix is invalid, and store this info in the (otherwise unused) most significant bit.

In another (yet unreleased) implementation, I'd like to decode [prosigns](https://en.wikipedia.org/wiki/Prosigns_for_Morse_code) and I figured out that I could look for their binary encoding (i.e. large values of `state`) in a separate data structure.

ghuysmans committed:
morsecode.c: Add early rejection test (#5)

ghuysmans committed:
morsecode.c: Restore early rejection

ghuysmans commented:
Hi Christopher,

I've implemented this idea and added a test to check whether `.-.-` is detected as an error before its terminator. The resulting x86-64 code is shorter than the original (when compiled using `gcc -O3`) and the data fit a single cache line, so I guess it might as well be faster while satisfying the desired properties.
